server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/http/AuthenticationRoutes.java

```diff
@@ -0,0 +1,270 @@
+    public AuthenticationRoutes(UsersRepository usersRepository, SimpleTokenManager simpleTokenManager, AccessTokenManager accessTokenManager, SimpleTokenFactory simpleTokenFactory, MetricFactory metricFactory, AuthenticationReactiveFilter authenticationReactiveFilter) {
+        this.mapper = new MultipleObjectMapperBuilder()
+            .registerClass(ContinuationTokenRequest.UNIQUE_JSON_PATH, ContinuationTokenRequest.class)
+            .registerClass(AccessTokenRequest.UNIQUE_JSON_PATH, AccessTokenRequest.class)
+            .build();
+        this.usersRepository = usersRepository;
+        this.simpleTokenManager = simpleTokenManager;
+        this.accessTokenManager = accessTokenManager;
+        this.simpleTokenFactory = simpleTokenFactory;
+        this.metricFactory = metricFactory;
+        this.authenticationReactiveFilter = authenticationReactiveFilter;
+    }
+
+    @Override
+    public Logger logger() {
+        return LOGGER;
+    }
+
+    @Override
+    public HttpServerRoutes define(HttpServerRoutes builder) {
+        return builder
+            .post(AUTHENTICATION, this::post)
+            .get(AUTHENTICATION, this::returnEndPointsResponse)
+            .delete(AUTHENTICATION, this::delete)
+            .options(AUTHENTICATION, CORS_CONTROL);
+    }
+
+    private Mono<Void> post(HttpServerRequest request, HttpServerResponse response) {
+        return Mono.from(metricFactory.runPublishingTimerMetricLogP99("JMAP-authentication-post",
+            Mono.just(request)
+                .map(this::assertJsonContentType)
+                .map(this::assertAcceptJsonOnly)
+                .flatMap(this::deserialize)
+                .flatMap(objectRequest -> {
+                    if (objectRequest instanceof ContinuationTokenRequest) {
+                        return handleContinuationTokenRequest((ContinuationTokenRequest) objectRequest, response);
+                    } else if (objectRequest instanceof AccessTokenRequest) {
+                        return handleAccessTokenRequest((AccessTokenRequest) objectRequest, response);
+                    } else {
+                        throw new RuntimeException(objectRequest.getClass() + " " + objectRequest);
+                    }
+                })
+                .onErrorResume(BadRequestException.class, e -> handleBadRequest(response, e))
+                .onErrorResume(e -> handleInternalError(response, e))))
+            .subscribeOn(Schedulers.elastic());
+    }
+
+    private Mono<Void> returnEndPointsResponse(HttpServerRequest req, HttpServerResponse resp) {
+        try {
+            return authenticationReactiveFilter.authenticate(req)
+                .then(resp.status(OK)
+                    .header(CONTENT_TYPE, JSON_CONTENT_TYPE_UTF8)
+                    .sendString(Mono.just(mapper.writeValueAsString(EndPointsResponse
+                        .builder()
+                        .api(JMAPUrls.JMAP)
+                        .eventSource("/notImplemented")
```

Comment:
That it's not implemented? :trollface: 

Joke aside, it was the case in the old code already. If you do compile the doc with Docker (you have a `doc` folder in `jmap-draft` containing the old spec, with a README explaining you how to run it locally) : http://localhost:8000/spec/authentication.html#201-authentication-is-complete-access-token-created

![screenshot-localhost_8000-2020 03 17-11_01_35](https://user-images.githubusercontent.com/9005025/76820836-d4763880-683e-11ea-9e58-20f23a103029.png)

Seems related to push, but was not defined yet when those specs got implemented, and this part completely changed since. It's still part of the specs implemented at that time, so I say we keep it, as it is. You ok with this?

Reply:
Should it be a constant in  JMAPUrls ?

Reply:
Could be, I'm not against it